### ISSUE TITLE
Normalized and Standardization of Version Numbers

### ISSUE THREAD

**Issue by hl2guide** [microsoft/winget-cli#1004]
# Description of the new feature/enhancement

Normalization of version number format to fix inconsistancies.

e.g.:

![example](https://user-images.githubusercontent.com/17044982/119868146-09f9d400-bf62-11eb-9e31-fa4173129ec8.jpg)

# Proposed technical implementation details (optional)

- Improvement of detecting actual installed version of a software package
- Cleaning of the version number to a standardized ruleset

**Comment by denelon** [comment]
This looks like a duplicate of https://github.com/microsoft/winget-cli/issues/981

**Comment by denelon** [comment]
@hl2guide the image you included is very helpful. If you don't mind, add it to the other issue and add your 👍. If you believe this is a duplicate, please close it. If not, let me know what you feel is different here.

**Comment by hl2guide** [comment]
@denelon I'd consider this seperate as that issue is more about a git formatted versioning,
where this issue is more about the smart handling of different versioning schemes (a reaction)
for substrings.

e.g.:

The normal ones [easier to handle and compare]

- 1.2
- 1.2.3
- 1.2.3.4

The non-standard ones [needs special handling]

- alpha, beta, etc
- 1.1a
- Build 10
- unknown (not determined)

**Comment by denelon** [comment]
Yes, we're aware of the challenge with the various versioning mechanisms with software. It would be pretty fantastic for ISVs/Publishers to adopt [semantic](https://semver.org) versioning. We are also working through best practices for packages that essentially get treated as different packages for "major" releases. Programming languages and compilers routinely fall into this category. We currently separate version strings into tokens and sort by each token in sequence. It's an attempt at treating versions semantically.

We also have some interesting challenges with "preview" versions of software with special strings for versions like these, and how to deal with software that can be installed "side by side" like Visual Studio Code, and software that cannot like Docker Desktop.

**Comment by leonardder** [comment]
Especially for software authors, it would be very helpful to know how winget calculates whether a package can be upgraded or not.
For example, the NVDA screen reader for the blind has a versioning scheme of year.major.minor, e.g. 2021.1.1, where the minor part is optional. The full versioning scheme is year.major.minor.build, e.g. 2021.1.1.22000. For alpha en beta versions, the display version is in the form alpha-22966,091b315e, whereas the file version is 2021.2.0.22966.
Winget yet treats version 2020.4 as an update of version alpha-22966,091b315e. Documentation could help us in:

1. Changing what is registered in the uninstall registry key to satisfy winget, if necessary
2. Finding out why alpha-22966,091b315e is considered older than 2020.4. Is this a bug or by design?

**Comment by hl2guide** [comment]
Updated example:

![example2](https://user-images.githubusercontent.com/17044982/123540129-bc4ed200-d780-11eb-9442-99019d5aabab.png)

**Comment by ItzLevvie** [comment]
f.lux and MEGAsync displaying "Unknown" is an issue that needs to be brought up to the applications' developers as the Windows Package Manager uses the values provided from the Control Panel. The only solution that the Windows Package Manager team can do here is to allow users to filter out versions that display as "Unknown" in the future or to solve it temporarily on the users' end they would've to add a `DisplayVersion` registry key in `Microsoft\Windows\CurrentVersion\Uninstall`.

Standard Notes may seem to be an issue with their installer as when I install the application using the Windows Package Manager, it shows up as `3.8.7` rather than `1.0` in Control Panel.

Unified Remote has now been temporarily removed from the repository, as it seems that there's a driver prompt blocking the install when ran from a clean installation of Windows.

VLC uses different version numbers between the MSI installers, which shows up as `3.0.16.0`, and the EXE installers, which shows up as `3.0.16`.

**Comment by denelon** [comment]
Revisiting this one. Since it was filed, the concrete and actionable parts of version handling have been implemented, so I'd like to close it as largely addressed.

**Semantic, tokenized version comparison.** WinGet parses each version into ordered parts, and each part is split into a numeric component plus an optional trailing string (e.g., `1.1a` -> `1` + `1a`). Trailing empty/zero parts are trimmed so `1.2.3.0`, `1.2.3`, and `1.2` compare as expected, and leading non-digit prefixes are handled. This is the "sort by each token in sequence" behavior described earlier in this thread. The closely related request #981 ("support versioning: major.minor.patch.int-somestring") has also since been completed.

**Publisher-side version reconciliation.** Manifests now support `AppsAndFeaturesEntries` with a `DisplayVersion`, which lets a publisher map the version registered in Apps & Features to the manifest version. This directly addresses the scenario @LeonarddeR raised (e.g., an installed `alpha-...` / file-version mismatch): a manifest can declare the `DisplayVersion` that the package actually registers so upgrade detection lines up. Details: [Add an installer to a manifest / AppsAndFeaturesEntries](https://learn.microsoft.com/windows/package-manager/package/manifest).

**What remains open-ended.** Automatically "cleaning" arbitrary version strings to a single standardized display ruleset - and robustly ordering freeform values like `alpha`, `beta`, `Build 10`, or `unknown` - isn't something WinGet can do reliably without publisher-provided context, precisely because those schemes carry no universal ordering. The mechanisms above give publishers the tools to make their versions comparable, which is the durable path forward.

If you hit a **specific** case where two concrete versions compare or sort incorrectly, please open a new issue with the exact installed version, the manifest version, and the expected result so we can treat it as a targeted bug. Closing this broader request as addressed.

_(Drafted with GitHub Copilot assistance.)_
